### PR TITLE
BUG: Fix permutation of distance matrices in scipy.stats.multiscale_graphcorr

### LINKED ISSUE #12155: BUG: Fix permutation of distance matrices in scipy.stats.multiscale_graphcorr
My issue is about ...

When permuting distance matrices in ``scipy.stats.multiscale_graphcorr``, the entire distance matrix is permuted instead of the rows and columns. This yields incorrect p-values in this case.

### PR DESCRIPTION
**Pull request by sampan501**
#### Reference issue

Closes #12155 

#### What does this implement/fix?

Properly permutes distance matrices, calculates the distance matrix first to save time, and adds more robust integration with ``scipy.spatial.cdist``.

### CODE REVIEW

**Comment by ilayn** [comment]
This adds more than a bug fix. Do you mind adding new functionality in a separate PR? Also note that we cannot change public API function signatures without deprecating a few versions hence some changes won't even fly.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -4434,11 +4430,11 @@ def multiscale_graphcorr(x, y, compute_distance=_euclidean_dist, reps=1000,
         shapes ``(n, p)`` and ``(m, p)``, this optional will be overriden and
         set to ``True``. Set to ``True`` if ``x`` and ``y`` both have shapes
         ``(n, p)`` and a two sample test is desired. The default is ``False``.
+        Note that this will not run if inputs are distance matrices.
     random_state : int or np.random.RandomState instance, optional
         If already a RandomState instance, use it.
         If seed is an int, return a new RandomState instance seeded with seed.
         If None, use np.random.RandomState. Default is None.
-
```
Comment (rgommers) [inline scipy/stats/stats.py:4441]: All of the blank line deletions in this diff are incorrect, they will break html rendering. Can you revert those?

**Comment by rgommers** [comment]
@sampan501 thanks for the fix. Could you please also add a regression test? It's now unclear what was broken before and works after the change.

File: scipy/stats/stats.py (hunk 1 above)
Comment (sampan501) [reply scipy/stats/stats.py]: Fixed

**Comment by sampan501** [comment]
@rgommers Does everything look good?

**Comment by rgommers** [comment]
@sampan501 the changes to the code look fine to me, however the test you added passes with scipy `1.4.1` - could you think of a better regression test? Or maybe just increasing test precision?

Also, could you add `@pytest.mark.slow`? These tests are painfully slow, ~5 seconds each.

**Comment by sampan501** [comment]
@rgommers Based on my testing, this new test does not pass in 1.4.1. It essentially validates that the p-value when calculating the p-value with distance matrices first vs. just the original data matrices is the same

**Comment by sampan501** [comment]
@rgommers Any feedback?

**Comment by sampan501** [comment]
@rgommers Just checking in. Any feedback?

**Review by rgommers — APPROVED** [review]
Tested now, LGTM. Thanks @sampan501, and apologies for the delay in reviewing - June was a bit crazy.

**Pull request merged**